### PR TITLE
Immutability using type provider with Scala 2.13

### PR DESCRIPTION
**Pull request by roterEmil**

### CODE REVIEW

File: DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala [hunk 1]
```diff
@@ -540,13 +538,16 @@ object Purity {
                 Console.println(usage)
                 return ;
         }
+        if (!fieldMutabilityAnalysisName.contains("L3")) {
```
Comment (errt) [inline DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala]: Add blank line before if

File: DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala [hunk 2]
```diff
@@ -540,13 +538,16 @@ object Purity {
...
-        if (eager) {
-            support ::= EagerClassImmutabilityAnalysis
-            support ::= EagerTypeImmutabilityAnalysis
-        } else {
-            support ::= LazyClassImmutabilityAnalysis
-            support ::= LazyTypeImmutabilityAnalysis
+                support ::= EagerClassImmutabilityAnalysis
+                support ::= EagerTypeImmutabilityAnalysis
+            } else {
+
```
Comment (errt) [inline DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala]: Remove these two blank lines

File: DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala (hunk 1 above)
Comment (errt) [inline DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala]: 1. Why is this a `contains` instead of an exact match?
2. Why would one not want to schedule Class and Type Immutability with the L3 analysis? Even if the L3FieldMutabilityAnalysis doesn't use them, the purity analyses still use them directly, at least according to their `uses` (L0PurityAnalysis uses only TypeImmutability, not ClassImmutability, though)
3. "L3" is not documented as a valid value in line 96
4. The variable and parameter seem to be ill-named, since they define the field assignability analysis.

File: DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala [hunk 3]
```diff
@@ -565,28 +566,26 @@ object Purity {
...
-                support ::= LazyUnsoundPrematurelyReadFieldsAnalysis
+                support ::= LazyL2FieldAssignabilityAnalysis
 
             case Some("none") =>
 
             case None => analysis match {
-                case LazyL0PurityAnalysis => LazyL0FieldMutabilityAnalysis
-                case LazyL1PurityAnalysis => LazyL1FieldMutabilityAnalysis
-                case LazyL2PurityAnalysis => LazyL1FieldMutabilityAnalysis
+                case LazyL0PurityAnalysis => LazyL0FieldAssignabilityAnalysis
```
Comment (errt) [inline DEVELOPING_OPAL/tools/src/main/scala/org/opalj/support/info/Purity.scala]: As IntelliJ correctly points out, this is broken, because the analyses are not actually added to `support`

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/DependentCalls.java [hunk 4]
```diff
@@ -42,24 +42,24 @@ public static DependentCalls createDependentCalls() {
...
-            eps = @EP(cf = DependentCalls.class, field = "myValue", pk = "FieldMutability",
-                    p = "EffectivelyFinalField"))
+            eps = @EP(cf = DependentCalls.class, field = "myValue", pk = "FieldImmutability",
+                    p = "TransitivelyImmutableField"))
     @Impure(value = "field used not recognized as effectively final",
-            eps = @EP(cf = DependentCalls.class, field = "myValue", pk = "FieldMutability",
-                    p = "EffectivelyFinalField"),
-            negate = true, analyses = L0PurityAnalysis.class)
+            eps = @EP(cf = DependentCalls.class, field = "myValue", pk = "FieldImmutability",
+                    p = "TransitivelyImmutableField"), negate=true,
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/DependentCalls.java]: Please restore formatting, the `negate = true` was on the next line on purpose (to have these two lines here only for the `@EP` specification for better readability)

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/ReferenceTypes.java [hunk 5]
```diff
@@ -64,24 +64,24 @@ public ReferenceTypes(ReferenceTypes other) {
     // Setting static fields is impure
 
     @Pure(value = "Uses final static immutable object",
-            eps = @EP(cf = DependentCalls.class, pk = "ClassImmutability", p = "ImmutableObject"))
+            eps = @EP(cf = DependentCalls.class, pk = "ClassImmutability", p = "TransitivelyImmutableClass"), analyses = {L1PurityAnalysis.class, L2PurityAnalysis.class})
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/ReferenceTypes.java]: It is unclear to me why this should now be necessary if it wasn't before.
If it really is, please move to a line of its own (see comment above).

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/ReferenceTypes.java [hunk 6]
```diff
@@ -64,24 +64,24 @@ public ReferenceTypes(ReferenceTypes other) {
...
+            eps = @EP(cf = DependentCalls.class, pk = "ClassImmutability", p = "TransitivelyImmutableClass"),
             negate = true, analyses = L0PurityAnalysis.class)
     public static DependentCalls getStaticFinalImmutableObj() {
         return staticFinalImmutableObj;
     }
 
     @Pure(value = "Uses effectively final static immutable object", eps = {
-            @EP(cf = DependentCalls.class, pk = "ClassImmutability", p = "ImmutableObject"),
+            @EP(cf = DependentCalls.class, pk = "ClassImmutability", p = "TransitivelyImmutableClass"),
             @EP(cf = ReferenceTypes.class, field = "staticEffectivelyFinalImmutableObj",
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/purity/ReferenceTypes.java]: Fields like this should be renamed to use the new terminology to avoid confusion.

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ImmutabilityTests_closedWorld.scala [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.opalj.tac.fpcf.analyses.immutability.field_assignability.EagerL2FieldAssignabilityAnalysis
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+
+/**
+ * Tests the immutability analysis with a closed world assumption.
+ *
+ * @author Tobias Roth
+ */
+class ImmutabilityTests_closedWorld extends PropertiesTest {
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ImmutabilityTests_closedWorld.scala]: I don't like this naming scheme. In Java/Scala underscores are not used, but CamelCase is.

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PropertiesTest.scala [hunk 8]
```diff
@@ -207,7 +207,7 @@ abstract class PropertiesTest extends AnyFunSpec with Matchers {
                     val nonFinalPSs = epss.filter(_.isRefinable)
                     assert(
                         nonFinalPSs.isEmpty,
-                        nonFinalPSs.mkString("some epss are not final:\n\t", "\n\t", "\n")
+                        nonFinalPSs.mkString("some eps are not final:\n\t", "\n\t", "\n")
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PropertiesTest.scala]: Not a typo, but the plural of `eps` (i.e., Entity-Property-StateS)

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala [hunk 9]
```diff
@@ -33,6 +33,8 @@ import org.opalj.tac.fpcf.analyses.purity.SystemOutLoggingAllExceptionRater
  */
 class PurityTests extends PropertiesTest {
 
+    import org.opalj.br.analyses.Project
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala]: Move to file scope imports

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala [hunk 10]
```diff
@@ -97,4 +102,29 @@ class PurityTests extends PropertiesTest {
         validateProperties(as, contextsWithAnnotations(as.project), Set("Purity"))
     }
 
-}
+    describe(
+        "the org.opalj.fpcf.analyses.L2PurityAnalysis is executed "+
+            "together with the L0FieldImmutabilityAnalysis and L2FieldAssignabilityAnalysis"
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala]: Is there enough of a conceptual difference between L1 and L2 field assignability? If the latter essentially subsumes the first, we should probably only test with that one?

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala [hunk 11]
```diff
@@ -97,4 +102,29 @@ class PurityTests extends PropertiesTest {
...
+    ) {
+
+            L2PurityAnalysis.setRater(Some(SystemOutLoggingAllExceptionRater))
+
+            val as = executeAnalyses(Set(
+                EagerL2PurityAnalysis,
+                LazyL0FieldImmutabilityAnalysis,
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/PurityTests.scala]: Please use the same ordering of analyses as for the other test cases to make it easier to compare the configurations.

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/ClassImmutabilityAnalysis.scala [hunk 12]
```diff
@@ -1,22 +1,28 @@
 /* BSD 2-Clause License - see OPAL/LICENSE for details. */
 package org.opalj
-package br
+package tac
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/ClassImmutabilityAnalysis.scala]: Why were this and the `TypeImmutabilityAnalysis` moved to `tac`? I don't see them using the TAC, so they should have remained in `br`.

File: OPAL/br/src/main/scala/org/opalj/br/fpcf/analyses/L0CompileTimeConstancyAnalysis.scala [hunk 13]
```diff
@@ -48,11 +51,12 @@ class L0CompileTimeConstancyAnalysis private[analyses] ( final val project: Some
...
         var dependee: EOptionP[Entity, Property] = {
-            propertyStore(field, FieldMutability.key) match {
-                case FinalP(LazyInitializedField) => return Result(field, CompileTimeVaryingField);
-                case FinalP(_: FinalField)        => return Result(field, CompileTimeConstantField);
-                case FinalP(_: NonFinalField)     => return Result(field, CompileTimeVaryingField);
-                case ep                           => ep
+            propertyStore(field, FieldAssignability.key) match {
+                case LBP(NonAssignable) => return Result(field, CompileTimeConstantField);
+                case UBP(EffectivelyNonAssignable |
+                    UnsafelyLazilyInitialized |
```
Comment (errt) [inline OPAL/br/src/main/scala/org/opalj/br/fpcf/analyses/L0CompileTimeConstancyAnalysis.scala]: Seems unnecessary to have on a line of its own

File: OPAL/br/src/main/scala/org/opalj/br/fpcf/analyses/L0CompileTimeConstancyAnalysis.scala [hunk 14]
```diff
@@ -48,11 +51,12 @@ class L0CompileTimeConstancyAnalysis private[analyses] ( final val project: Some
...
 
         var dependee: EOptionP[Entity, Property] = {
-            propertyStore(field, FieldMutability.key) match {
-                case FinalP(LazyInitializedField) => return Result(field, CompileTimeVaryingField);
-                case FinalP(_: FinalField)        => return Result(field, CompileTimeConstantField);
-                case FinalP(_: NonFinalField)     => return Result(field, CompileTimeVaryingField);
-                case ep                           => ep
+            propertyStore(field, FieldAssignability.key) match {
+                case LBP(NonAssignable) => return Result(field, CompileTimeConstantField);
+                case UBP(EffectivelyNonAssignable |
```
Comment (errt) [inline OPAL/br/src/main/scala/org/opalj/br/fpcf/analyses/L0CompileTimeConstancyAnalysis.scala]: Why is effective non-assignability here and not grouped with full non-assignability?

File: OPAL/framework/src/main/resources/ApplicationProject.conf [hunk 15]
```diff
@@ -0,0 +1,21 @@
+org.opalj.br.analyses {
```
Comment (errt) [inline OPAL/framework/src/main/resources/ApplicationProject.conf]: This already exists identically as `CommandLineProject.conf` in the same directory. Why was this added?

File: OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala [hunk 16]
```diff
@@ -93,6 +93,7 @@ trait Expr[+V <: Var[V]] extends ASTNode[V] {
     def asMethodTypeConst: MethodTypeConst = throw new ClassCastException();
     def isMethodHandleConst: Boolean = false
     def asMethodHandleConst: MethodHandleConst = throw new ClassCastException();
+    def isCompare: Boolean = false
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala]: Put next to `asCompare`

File: OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala [hunk 17]
```diff
@@ -127,6 +129,7 @@ trait Expr[+V <: Var[V]] extends ASTNode[V] {
     def asGetStatic: GetStatic = throw new ClassCastException();
     def asInvokedynamicFunctionCall: InvokedynamicFunctionCall[V] = throw new ClassCastException();
     def asFunctionCall: FunctionCall[V] = throw new ClassCastException();
+    def isFunctionCall: Boolean = false
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala:130]: Put above `asFunctionCall`, as the other `as`/`is` combinations are ordered that way as well.

File: OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala [hunk 18]
```diff
@@ -560,6 +563,7 @@ object NewArray { final val ASTID = -18 }
 case class ArrayLoad[+V <: Var[V]](pc: PC, index: Expr[V], arrayRef: Expr[V]) extends ArrayExpr[V] {
 
     final override def asArrayLoad: this.type = this
+    final override def isArrayLoad: Boolean = true
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala:564]: Put before the `as` method

File: OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala [hunk 19]
```diff
@@ -782,6 +786,7 @@ sealed abstract class FunctionCall[+V <: Var[V]] extends Expr[V] with Call[V] {
     final override def isValueExpression: Boolean = false
     final override def isVar: Boolean = false
     final override def asFunctionCall: this.type = this
+    final override def isFunctionCall: Boolean = true
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Expr.scala:787]: Put before the `as` method

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 20]
```diff
@@ -65,10 +65,12 @@ sealed abstract class Stmt[+V <: Var[V]] extends ASTNode[V] {
     def asMonitorExit: MonitorExit[V] = throw new ClassCastException();
     def asArrayStore: ArrayStore[V] = throw new ClassCastException();
     def asThrow: Throw[V] = throw new ClassCastException();
+    def isThrow: Boolean = false
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Group with the other `is` methods below

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 21]
```diff
@@ -65,10 +65,12 @@ sealed abstract class Stmt[+V <: Var[V]] extends ASTNode[V] {
...
     def asFieldWriteAccessStmt: FieldWriteAccessStmt[V] = throw new ClassCastException();
     def asPutStatic: PutStatic[V] = throw new ClassCastException();
     def asPutField: PutField[V] = throw new ClassCastException();
     /*inner type*/ def asMethodCall: MethodCall[V] = throw new ClassCastException();
+    def isMethodCall: Boolean = false
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Group with the other `is` methods below

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 22]
```diff
@@ -77,14 +79,23 @@ sealed abstract class Stmt[+V <: Var[V]] extends ASTNode[V] {
     def asExprStmt: ExprStmt[V] = throw new ClassCastException();
     def asCaughtException: CaughtException[V] = throw new ClassCastException();
     def asCheckcast: Checkcast[V] = throw new ClassCastException();
+    def isCheckcast: Boolean = false;
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Group with the other `is` methods below

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 23]
```diff
@@ -77,14 +79,23 @@ sealed abstract class Stmt[+V <: Var[V]] extends ASTNode[V] {
...
 
     def isAssignment: Boolean = false
+    def isArrayStore: Boolean = false
     def isExprStmt: Boolean = false
     def isNonVirtualMethodCall: Boolean = false
     def isVirtualMethodCall: Boolean = false
     def isStaticMethodCall: Boolean = false
     def isCaughtException: Boolean = false
 
+    def isIf: Boolean = false
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Why is this offset with a blank line? Remove and sort these methods the same way the `as` methods are sorted above.

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 24]
```diff
@@ -504,6 +522,9 @@ sealed abstract class SynchronizationStmt[+V <: Var[V]] extends Stmt[V] {
 case class MonitorEnter[+V <: Var[V]](pc: PC, objRef: Expr[V]) extends SynchronizationStmt[V] {
 
     final override def asMonitorEnter: this.type = this
+
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Remove blank lines

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 25]
```diff
@@ -530,6 +551,8 @@ object MonitorEnter {
 case class MonitorExit[+V <: Var[V]](pc: PC, objRef: Expr[V]) extends SynchronizationStmt[V] {
 
     final override def asMonitorExit: this.type = this
+
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Remove blank line

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 26]
```diff
@@ -695,14 +721,14 @@ case class PutField[+V <: Var[V]](
         objRef:            Expr[V],
         value:             Expr[V]
 ) extends FieldWriteAccessStmt[V] {
-
+    final override def isPutField: Boolean = true
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala]: Put below the `as` method

File: OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala [hunk 27]
```diff
@@ -742,6 +768,8 @@ sealed abstract class MethodCall[+V <: Var[V]] extends Stmt[V] with Call[V] {
 
     final override def asMethodCall: this.type = this
 
+    final override def isMethodCall: Boolean = true
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/Stmt.scala:764]: Remove blank lines

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/L0FieldImmutabilityAnalysis.scala [hunk 28]
```diff
@@ -0,0 +1,429 @@
...
+import org.opalj.br.ConstantDouble
+import org.opalj.br.Synthetic
+import org.opalj.br.RuntimeVisibleAnnotationTable
+import org.opalj.br.Deprecated
+
+/**
+ * Analysis that determines the immutability of org.opalj.br.Field
+ * @author Tobias Roth
+ */
+class L0FieldImmutabilityAnalysis private[analyses] (val project: SomeProject)
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/L0FieldImmutabilityAnalysis.scala]: Why is this analysis called `L0`? Is there any reason to ever expect there to be a more precise variant of this analysis? Since it's just an aggregating analysis, I would have assumed this to be the definitive implementation and, thus, not have a level specifier.

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/field_assignability/AbstractFieldAssignabilityAnalysis.scala [hunk 29]
```diff
@@ -0,0 +1,324 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj
+package tac
+package fpcf
+package analyses
+package immutability
+package field_assignability
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/immutability/field_assignability/AbstractFieldAssignabilityAnalysis.scala]: I don't like this package name. Avoid underscores

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/AbstractPurityAnalysis.scala [hunk 30]
```diff
@@ -638,7 +642,7 @@ trait AbstractPurityAnalysis extends FPCFAnalysis {
     }
 
     def resolveDomainSpecificRater(fqn: String): DomainSpecificRater = {
-        import scala.reflect.runtime.universe.runtimeMirror
+
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/AbstractPurityAnalysis.scala:645]: Why was this moved? It is a very local use after all

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/L1PurityAnalysis.scala [hunk 31]
```diff
@@ -278,12 +282,14 @@ class L1PurityAnalysis private[analyses] (val project: SomeProject) extends Abst
                     return Result(state.context, ImpureByAnalysis)
 
             // Cases that are pure
-            case FinalP(_: FinalField)                   => // Reading eff. final fields
-            case FinalP(ImmutableType | ImmutableObject) => // Returning immutable reference
+            case FinalP(NonAssignable | EffectivelyNonAssignable | LazilyInitialized) => // Reading eff. final fields
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/L1PurityAnalysis.scala]: Adapt comments to new terminology

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/L1PurityAnalysis.scala [hunk 32]
```diff
@@ -278,12 +282,14 @@ class L1PurityAnalysis private[analyses] (val project: SomeProject) extends Abst
...
+            case FinalP(TransitivelyImmutableType |
+                TransitivelyImmutableClass) => // Returning immutable reference
 
             // Cases resulting in side-effect freeness
-            case FinalP(_: FieldMutability | // Reading non-final field
-                _: TypeImmutability | _: ClassImmutability) => // Returning mutable reference
+            case FinalP(_: FieldAssignability | // Reading non-final field
+                _: TypeImmutability |
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/purity/L1PurityAnalysis.scala]: Probably not necessary to put everything on a line of its own, especially since the comment referens to both, class and type immutability

**Review by errt — CHANGES_REQUESTED** [review]
Partial review, mostly purity related classes and a few important framework classes

**Review by TorunR — DISMISSED** [review]

**Review by errt — APPROVED** [review]
